presto-hive/src/main/java/com/facebook/presto/hive/util/TempFileWriter.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.facebook.presto.orc.OrcWriter;
+import com.facebook.presto.orc.OrcWriterOptions;
+import com.facebook.presto.orc.OrcWriterStats;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableMap;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_CLOSE_ERROR;
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_DATA_ERROR;
+import static com.facebook.presto.orc.OrcEncoding.ORC;
+import static com.facebook.presto.orc.metadata.CompressionKind.LZ4;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.joda.time.DateTimeZone.UTC;
+
+public class TempFileWriter
+        implements Closeable
+{
+    private final PageBuilder pageBuilder;
+    private final OrcWriter orcWriter;
+
+    public TempFileWriter(List<Type> types, OutputStream output)
+    {
+        this.pageBuilder = new PageBuilder(types);
+        this.orcWriter = createOrcFileWriter(output, types);
+    }
+
+    public void writePages(List<Page> pages, int[] pageIndexes, int[] positionIndexes)
```

Comment:
move this to `SortBuffer`.  This is the only implementer and there is no reason for this class to deal with the complexity... also it will make the short circuit write when data is small super easy.